Strictly enforce exclusivity for the “details” element “name” attribute

sideshowbarker opened the pull request:
#### 672fe76f367734e84b2bad012b9676281f813d71
<pre>
Strictly enforce exclusivity for the “details” element “name” attribute
<a href="https://bugs.webkit.org/show_bug.cgi?id=262187">https://bugs.webkit.org/show_bug.cgi?id=262187</a>

Reviewed by Darin Adler.

This adds the necessary changes to bring WebKit into full conformance
with the spec change in <a href="https://github.com/whatwg/html/commit/e0cf318.">https://github.com/whatwg/html/commit/e0cf318.</a>
In particular, this commit implements the requirements of the “ensure
details exclusivity by closing the given element if needed” algorithm at
<a href="https://html.spec.whatwg.org/#ensure-details-exclusivity-by-closing-the-given-element-if-needed">https://html.spec.whatwg.org/#ensure-details-exclusivity-by-closing-the-given-element-if-needed</a>

* LayoutTests/imported/w3c/web-platform-tests/html/semantics/interactive-elements/the-details-element/name-attribute-expected.txt:
Renamed from LayoutTests/imported/w3c/web-platform-tests/html/semantics/interactive-elements/the-details-element/name-attribute.tentative-expected.txt.
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/interactive-elements/the-details-element/name-attribute.html:
Renamed from LayoutTests/imported/w3c/web-platform-tests/html/semantics/interactive-elements/the-details-element/name-attribute.tentative.html.

* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
Add ShouldNotFireMutationEventsScope.h

* Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
(WebCore::AccessibilityNodeObject::setIsExpanded):
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::isExpanded const):

* Source/WebCore/dom/ContainerNode.cpp:
(WebCore::ContainerNode::removeNodeWithScriptAssertion):
Only call dispatchChildRemovalEvents() if document.shouldNotFireMutationEvents() is false.
* Source/WebCore/dom/Document.h:
(WebCore::Document::shouldNotFireMutationEvents const):
(WebCore::Document::setShouldNotFireMutationEvents):
Added shouldNotFireMutationEvents() and setShouldNotFireMutationEvents()
* Source/WebCore/dom/Node.cpp:
(WebCore::Node::dispatchSubtreeModifiedEvent):
Return early from dispatchSubtreeModifiedEvent() if document.shouldNotFireMutationEvents() is true.

* Source/WebCore/dom/ShouldNotFireMutationEventsScope.h: Added.
(WebCore::ShouldNotFireMutationEventsScope::ShouldNotFireMutationEventsScope):
(WebCore::ShouldNotFireMutationEventsScope::~ShouldNotFireMutationEventsScope):

* Source/WebCore/html/HTMLDetailsElement.cpp:
(WebCore::HTMLDetailsElement::didAddUserAgentShadowRoot):
(WebCore::HTMLDetailsElement::attributeChanged):
(WebCore::HTMLDetailsElement::insertedIntoAncestor):
(WebCore::HTMLDetailsElement::otherElementsInNameGroup):
(WebCore::HTMLDetailsElement::ensureDetailsExclusivityAfterMutation):
(WebCore::HTMLDetailsElement::toggleOpen):
* Source/WebCore/html/HTMLDetailsElement.h:

Canonical link: <a href="https://commits.webkit.org/269127@main">https://commits.webkit.org/269127@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d59794300a596ac035558b63b73802efe04f6d5f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/21563 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/26/builds/21868 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/22616 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/23428 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/19980 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/21806 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/25242 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/22115 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/21166 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/21790 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/21451 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/18696 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/24279 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/18651 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/19546 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/25848 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/19710 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/19753 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/23698 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/20236 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/17244 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/19556 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5168 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/23792 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/20139 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/dom/Document.h:
```diff
@@ -2270,6 +2273,8 @@ class Document
 
     MediaProducerMediaStateFlags m_mediaState;
 
+    bool m_fireMutationEvents = true;
```
Boolean variables need a prefix in WebKit. This should be `m_shouldFireMutationEvents` and the setter `setShouldFireMutationEvent()`. Your getter is already correct.

sideshowbarker replied:
> Boolean variables need a prefix in WebKit. This should be `m_shouldFireMutationEvents` and the setter `setShouldFireMutationEvent()`.

Thanks — made those changes

darinadler commented on Source/WebCore/dom/Document.h:
```diff
@@ -441,6 +441,9 @@ class Document
 
     bool canContainRangeEndPoint() const final { return true; }
 
+    bool shouldFireMutationEvents() const { return m_shouldFireMutationEvents; }
```
It seems unfortunate that we have to create this new mechanism, particularly a global one for the entire document. Is there some way we can do this with something we already have? There are already so many ways we do things without firing mutation events, such as the "parser" mode for adding children. Is this really the first time we had to manipulate a document without firing mutation events?

sideshowbarker replied:
> Is this really the first time we had to manipulate a document without firing mutation events?

I had wondered about the same thing — and checked before writing up the patch, but didn’t find anything.

And it did feel some hesitation about needing to add it globally — but at https://html.spec.whatwg.org/multipage/dom.html#concept-document-fire-mutation-events-flag that spec says:

> Each [`Document`](https://html.spec.whatwg.org/multipage/dom.html#document) has a fire mutation events flag, which is a boolean, initially true.

That is, the spec explicitly says it’s a flag on `Document`. And the existing WPT tests do include cases for checking that the effect is global.

Still, I’d hoped a clever alternative for implementing it without hanging more onto `Document` might emerge during review. I messed around with trying some things locally, but I couldn’t make anything else work.

darinadler commented on Source/WebCore/html/HTMLDetailsElement.cpp:
```diff
@@ -160,23 +160,53 @@ void HTMLDetailsElement::attributeChanged(const QualifiedName& name, const AtomS
+    if (auto& name = attributeWithoutSynchronization(nameAttr); m_isOpen && !name.isEmpty() && document().settings().detailsNameAttributeEnabled()) {
+        bool shouldFireMutationEvents = document().shouldFireMutationEvents();
+        document().setShouldFireMutationEvents(false);
+        for (RefPtr otherDetailsElement : otherElementsInNameGroup()) {
```
Using `auto&` here would avoid a little reference count churn by not copying every `RefPtr` out of the vector.

sideshowbarker replied:
✅ Changed

darinadler commented on Source/WebCore/html/HTMLDetailsElement.cpp:
```diff
@@ -160,23 +160,53 @@ void HTMLDetailsElement::attributeChanged(const QualifiedName& name, const AtomS
 
+void HTMLDetailsElement::ensureDetailsExclusivityAfterMutation()
+{
+    if (auto& name = attributeWithoutSynchronization(nameAttr); m_isOpen && !name.isEmpty() && document().settings().detailsNameAttributeEnabled()) {
```
Seems slightly wasteful to do the hash table lookup for the `name` attribute even if `m_isOpen` is false.

sideshowbarker replied:
> Seems slightly wasteful to do the hash table lookup for the `name` attribute even if `m_isOpen` is false.

I guess this is moot now that that `m_isOpen`  is gone? Anyway, I did otherwise switch the order such that `document().settings().detailsNameAttributeEnabled()` preference check is called before the lookups for the `name` attribute and (now `open` attribute) — because that’s how most of the existing code for similar cases seems to order the checks. (And I guess maybe in general a preference check is less expensive than an attribute lookup?)

darinadler commented on Source/WebCore/html/HTMLDetailsElement.cpp:
```diff
@@ -160,23 +160,53 @@ void HTMLDetailsElement::attributeChanged(const QualifiedName& name, const AtomS
+Vector<RefPtr<HTMLDetailsElement>> HTMLDetailsElement::otherElementsInNameGroup()
+{
+    Vector<RefPtr<HTMLDetailsElement>> otherElementsInNameGroup;
+    for (auto& detailsElement : descendantsOfType<HTMLDetailsElement>(rootNode())) {
```
From a code readability standpoint, I think the single word local variable name `element` would be better than `detailsElement`.

sideshowbarker replied:
✅ Changed

darinadler commented on Source/WebCore/dom/Document.h:
```diff
@@ -2270,6 +2273,8 @@ class Document
 
     MediaProducerMediaStateFlags m_mediaState;
 
+    bool m_shouldFireMutationEvents = true;
```
WebKit code uses the `{ true }` syntax.

sideshowbarker replied:
✅ Changed

darinadler commented on Source/WebCore/html/HTMLDetailsElement.cpp:
```diff
@@ -160,23 +160,53 @@ void HTMLDetailsElement::attributeChanged(const QualifiedName& name, const AtomS
-                            otherElementsInNameGroup.append(&detailsElement);
-                    }
-                    for (RefPtr otherDetailsElement : otherElementsInNameGroup)
+                    bool shouldFireMutationEvents = document().shouldFireMutationEvents();
```
In WebKit most other things like this use RAII instead of requiring the caller to stash and restore a value. An example of that idiom is `ScriptDisallowedScope`.

sideshowbarker replied:
> In WebKit most other things like this use RAII instead of requiring the caller to stash and restore a value. An example of that idiom is `ScriptDisallowedScope`.

✅ Thanks — now added a `ShouldNotFireMutationEventsScope` for handling that.

darinadler commented on Source/WebCore/html/HTMLDetailsElement.cpp:
```diff
@@ -160,23 +160,53 @@ void HTMLDetailsElement::attributeChanged(const QualifiedName& name, const AtomS
-                    for (RefPtr otherDetailsElement : otherElementsInNameGroup)
+                    bool shouldFireMutationEvents = document().shouldFireMutationEvents();
+                    document().setShouldFireMutationEvents(false);
+                    for (RefPtr otherDetailsElement : otherElementsInNameGroup())
```
Using `auto&` here would avoid a little reference count churn by not copying every `RefPtr` out of the vector.

sideshowbarker replied:
✅ Changed

darinadler commented on Source/WebCore/html/HTMLDetailsElement.cpp:
```diff
@@ -160,23 +160,53 @@ void HTMLDetailsElement::attributeChanged(const QualifiedName& name, const AtomS
+        bool shouldFireMutationEvents = document().shouldFireMutationEvents();
+        document().setShouldFireMutationEvents(false);
+        for (RefPtr otherDetailsElement : otherElementsInNameGroup()) {
+            if (otherDetailsElement->m_isOpen) {
```
It’s not a great pattern that the `details` element caches the value of the `open` attribute with `m_isOpen`. There’s no need to store that in a data member for correct behavior and I doubt that it’s an important performance optimization.

sideshowbarker replied:
> It’s not a great pattern that the `details` element caches the value of the `open` attribute with `m_isOpen`. There’s no need to store that in a data member for correct behavior and I doubt that it’s an important performance optimization.

Thanks — it’s gone now. (And I’m glad to see it go — it had been part of the implementation since before I patched it, and I’d wondered myself whether it could actually be providing any win. And after now looking back at the commit history, I can see it seems to have been added very early on in the implementation — rather than being, say, some optimization that was intentionally later.)

darinadler commented on Source/WebCore/html/HTMLDetailsElement.cpp:
```diff
@@ -160,23 +160,53 @@ void HTMLDetailsElement::attributeChanged(const QualifiedName& name, const AtomS
+        document().setShouldFireMutationEvents(false);
+        for (RefPtr otherDetailsElement : otherElementsInNameGroup()) {
+            if (otherDetailsElement->m_isOpen) {
+                toggleOpen();
```
This seems fragile to me. Are we guaranteed that mutation events are the only way `toggleOpen` can lead to arbitrary execution of JavaScript? Because if anything runs and then in turn does other DOM modification, it will run with mutation events disabled. Calling `setShouldFireMutationEvents(false)` alone seems like a not entirely robust guarantee that no events of any kind will be dispatched. I’m sure that’s true right now, but I can so easily imagine us making a mistake later as we evolve the implementation.

sideshowbarker replied:
> Are we guaranteed that mutation events are the only way `toggleOpen` can lead to arbitrary execution of JavaScript?

I guess we probably do not have such a guarantee.

> Calling `setShouldFireMutationEvents(false)` alone seems like a not entirely robust guarantee that no events of any kind will be dispatched. I’m sure that’s true right now, but I can so easily imagine us making a mistake later as we evolve the implementation.

I understand but I can’t myself at this point think of an alternative for handling it more robustly, and I’m not sure how to proceed as far as getting proper resolution on this review comment.

For now at least, I’ve opened https://bugs.webkit.org/show_bug.cgi?id=262892. But I don’t know if raising a bug like that for this issue is in line with project norms; if not, I’m happy to close that bug, and instead try to get a further change made here before this gets merged.

darinadler commented on Source/WebCore/dom/ShouldNotFireMutationEventsScope.h:
```diff
@@ -0,0 +1,46 @@
+    ShouldNotFireMutationEventsScope(Document& document)
+        : m_document(document)
+    {
+        document.setShouldNotFireMutationEvents(true);
```
Need to assert here that it’s `false`, or save the old value.

sideshowbarker replied:
Thanks — added assert

darinadler commented on Source/WebCore/html/HTMLDetailsElement.cpp:
```diff
@@ -115,7 +116,7 @@ void HTMLDetailsElement::didAddUserAgentShadowRoot(ShadowRoot& root)
 
     m_defaultSlot = HTMLSlotElement::create(slotTag, document());
-    ASSERT(!m_isOpen);
+    ASSERT(attributeWithoutSynchronization(openAttr).isNull());
```
`hasAttribute` is more efficient and does the same thing

sideshowbarker replied:
Thanks — switched all

darinadler commented on Source/WebCore/html/HTMLDetailsElement.cpp:
```diff
@@ -152,36 +153,60 @@ void HTMLDetailsElement::queueDetailsToggleEventTask(DetailsState oldState, Deta
 void HTMLDetailsElement::toggleOpen()
 {
-    setBooleanAttribute(openAttr, !m_isOpen);
+    setBooleanAttribute(openAttr, attributeWithoutSynchronization(openAttr).isNull() ? true : false);
```
`hasAttribute`

darinadler commented on Source/WebCore/html/HTMLDetailsElement.cpp:
```diff
@@ -152,36 +153,60 @@ void HTMLDetailsElement::queueDetailsToggleEventTask(DetailsState oldState, Deta
+    if (document().settings().detailsNameAttributeEnabled() && !attributeWithoutSynchronization(openAttr).isNull() && !attributeWithoutSynchronization(nameAttr).isEmpty()) {
+        ShouldNotFireMutationEventsScope scope(document());
+        for (auto& otherDetailsElement : otherElementsInNameGroup()) {
+            if (!otherDetailsElement->attributeWithoutSynchronization(openAttr).isNull()) {
```
`hasAttribute`

darinadler commented on Source/WebCore/html/HTMLDetailsElement.cpp:
```diff
@@ -152,36 +153,60 @@ void HTMLDetailsElement::queueDetailsToggleEventTask(DetailsState oldState, Deta
 
+void HTMLDetailsElement::ensureDetailsExclusivityAfterMutation()
+{
+    if (document().settings().detailsNameAttributeEnabled() && !attributeWithoutSynchronization(openAttr).isNull() && !attributeWithoutSynchronization(nameAttr).isEmpty()) {
```
`hasAttribute` for `open`

darinadler commented on Source/WebCore/html/HTMLDetailsElement.cpp:
```diff
@@ -152,36 +153,60 @@ void HTMLDetailsElement::queueDetailsToggleEventTask(DetailsState oldState, Deta
 void HTMLDetailsElement::toggleOpen()
 {
-    setBooleanAttribute(openAttr, !m_isOpen);
+    setBooleanAttribute(openAttr, hasAttribute(openAttr) ? false : true);
```
I think `!hasAttribute(openAttr)` is more straightforward.

sideshowbarker replied:
OK — made it so

cdumez commented on Source/WebCore/dom/ShouldNotFireMutationEventsScope.h:
```diff
@@ -0,0 +1,48 @@
+    }
+
+private:
+    Document& m_document;
```
This needs to be a `CheckedRef<Document> m_document;`

sideshowbarker replied:
Thanks — I had actually read https://github.com/WebKit/WebKit/wiki/Smart-Pointer-Usage-Guidelines a while back, but forgotten in the meantime…

cdumez commented on Source/WebCore/html/HTMLDetailsElement.cpp:
```diff
@@ -152,36 +153,60 @@ void HTMLDetailsElement::queueDetailsToggleEventTask(DetailsState oldState, Deta
 void HTMLDetailsElement::toggleOpen()
 {
-    setBooleanAttribute(openAttr, !m_isOpen);
+    setBooleanAttribute(openAttr, !hasAttribute(openAttr) ? true : false);
```
`setBooleanAttribute(openAttr, !hasAttribute(openAttr));`

sideshowbarker replied:
✅ Thanks

sideshowbarker committed:
Strictly enforce exclusivity for the “details” element “name” attribute
https://bugs.webkit.org/show_bug.cgi?id=262187

Reviewed by Darin Adler.

This adds the necessary changes to bring WebKit into full conformance
with the spec change in https://github.com/whatwg/html/commit/e0cf318.
In particular, this commit implements the requirements of the “ensure
details exclusivity by closing the given element if needed” algorithm at
https://html.spec.whatwg.org/#ensure-details-exclusivity-by-closing-the-given-element-if-needed

* LayoutTests/imported/w3c/web-platform-tests/html/semantics/interactive-elements/the-details-element/name-attribute-expected.txt:
Renamed from LayoutTests/imported/w3c/web-platform-tests/html/semantics/interactive-elements/the-details-element/name-attribute.tentative-expected.txt.
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/interactive-elements/the-details-element/name-attribute.html:
Renamed from LayoutTests/imported/w3c/web-platform-tests/html/semantics/interactive-elements/the-details-element/name-attribute.tentative.html.

* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
Add ShouldNotFireMutationEventsScope.h

* Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
(WebCore::AccessibilityNodeObject::setIsExpanded):
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::isExpanded const):

* Source/WebCore/dom/ContainerNode.cpp:
(WebCore::ContainerNode::removeNodeWithScriptAssertion):
Only call dispatchChildRemovalEvents() if document.shouldNotFireMutationEvents() is false.
* Source/WebCore/dom/Document.h:
(WebCore::Document::shouldNotFireMutationEvents const):
(WebCore::Document::setShouldNotFireMutationEvents):
Added shouldNotFireMutationEvents() and setShouldNotFireMutationEvents()
* Source/WebCore/dom/Node.cpp:
(WebCore::Node::dispatchSubtreeModifiedEvent):
Return early from dispatchSubtreeModifiedEvent() if document.shouldNotFireMutationEvents() is true.

* Source/WebCore/dom/ShouldNotFireMutationEventsScope.h: Added.
(WebCore::ShouldNotFireMutationEventsScope::ShouldNotFireMutationEventsScope):
(WebCore::ShouldNotFireMutationEventsScope::~ShouldNotFireMutationEventsScope):

* Source/WebCore/html/HTMLDetailsElement.cpp:
(WebCore::HTMLDetailsElement::didAddUserAgentShadowRoot):
(WebCore::HTMLDetailsElement::attributeChanged):
(WebCore::HTMLDetailsElement::insertedIntoAncestor):
(WebCore::HTMLDetailsElement::otherElementsInNameGroup):
(WebCore::HTMLDetailsElement::ensureDetailsExclusivityAfterMutation):
(WebCore::HTMLDetailsElement::toggleOpen):
* Source/WebCore/html/HTMLDetailsElement.h:

Canonical link: https://commits.webkit.org/269127@main

rniwa commented on Source/WebCore/dom/ContainerNode.cpp:
```diff
@@ -191,10 +191,12 @@ ALWAYS_INLINE bool ContainerNode::removeNodeWithScriptAssertion(Node& childToRem
         ChildListMutationScope(*this).willRemoveChild(childToRemove);
     }
 
-    ASSERT_WITH_SECURITY_IMPLICATION(ScriptDisallowedScope::InMainThread::isEventDispatchAllowedInSubtree(childToRemove));
```
This change is not great. Now we're not asserting that script execution is allowed whenever parser invokes this function. Also, we didn't update ContainerNode::removeAllChildrenWithScriptAssertion.

sideshowbarker replied:
OK — I can put together a PR with a patch to fix those problems

sideshowbarker replied:
> Also, we didn't update ContainerNode::removeAllChildrenWithScriptAssertion.

What change is needed to ContainerNode::removeAllChildrenWithScriptAssertion?

rniwa replied:
It should be checking the flag just like ContainerNode::removeNodeWithScriptAssertion

rniwa replied:
Also, instead of suppressing assertion like this, we should do the closing of other details element in `didFinishInsertingNode` instead like this:
```
diff --git a/Source/WebCore/html/HTMLDetailsElement.cpp b/Source/WebCore/html/HTMLDetailsElement.cpp
index c45ff67165dd..2c6b978eb524 100644
--- a/Source/WebCore/html/HTMLDetailsElement.cpp
+++ b/Source/WebCore/html/HTMLDetailsElement.cpp
@@ -150,6 +150,7 @@ void HTMLDetailsElement::queueDetailsToggleEventTask(DetailsState oldState, Deta
 Node::InsertedIntoAncestorResult HTMLDetailsElement::insertedIntoAncestor(InsertionType insertionType, ContainerNode& parentOfInsertedTree)
+{
+    HTMLElement::insertedIntoAncestor(insertionType, parentOfInsertedTree);
+    return InsertedIntoAncestorResult::NeedsPostInsertionCallback;
+}
+
+void HTMLDetailsElement::didFinishInsertingNode()
 {
     ensureDetailsExclusivityAfterMutation();
-    return HTMLElement::insertedIntoAncestor(insertionType, parentOfInsertedTree);
 }
 
```

rniwa replied:
It's definitely not okay / safe to invoke `AXObjectCache::postNotification` inside `insertedIntoAncestor`.

rniwa replied:
Addressing this and other issues in https://github.com/WebKit/WebKit/pull/25095.

sideshowbarker replied:
> Also, instead of suppressing assertion like this, we should do the closing of other details element in `didFinishInsertingNode` instead like this: …

Thanks — I’d never yet used `didFinishInsertingNode` in any of the patches I’ve worked on. And walking through it in the code now I see:

https://github.com/WebKit/WebKit/blob/a2066e9a79c56b8a83d98ef94bf6b818cea23eb1/Source/WebCore/dom/Node.h#L475-L478

That is, especially the _“insertedInto **MUST NOT** invoke scripts. Return NeedsPostInsertionCallback and implement didFinishInsertingNode instead to run scripts”_ comment.

So it’s clear to me now (and for if/when the next time comes that I need to do something like this in a later patch for another feature like this) how cases like this should be handled.

rniwa commented on Source/WebCore/dom/ContainerNode.cpp:
```diff
@@ -191,10 +191,12 @@ ALWAYS_INLINE bool ContainerNode::removeNodeWithScriptAssertion(Node& childToRem
         childToRemove.notifyMutationObserversNodeWillDetach();
-        dispatchChildRemovalEvents(protectedChildToRemove);
+        if (!document().shouldNotFireMutationEvents()) {
+            RELEASE_ASSERT_WITH_SECURITY_IMPLICATION(ScriptDisallowedScope::InMainThread::isEventDispatchAllowedInSubtree(childToRemove));
```
This release assert is useless given isEventDispatchAllowedInSubtree() trivially returns true when `ASSERT_ENABLED || ENABLE(SECURITY_ASSERTIONS)` is false. We should have kept it as a debug assert instead.

rniwa replied:
Addressing this and other issues in https://github.com/WebKit/WebKit/pull/25095

sideshowbarker replied:
(Meta: The thumbs-up on the comments here are just ACKs from me that I read and understood the comments…)
